packages/compiler/src/install/install.ts

```diff
@@ -0,0 +1,197 @@
+}
+
+async function resolvePackageManagerSpecOrFail(
+  host: CliCompilerHost,
+  tracer: Tracer,
+  options: InstallTypeSpecDependenciesOptions,
+): Promise<[ResolvedSpecResult, readonly Diagnostic[]]> {
+  const result = await resolvePackageManagerSpec(host, tracer, options.directory);
+
+  switch (result.kind) {
+    case "no-package":
+      throw new InstallDependenciesError("No package.json found, cannot install dependencies.");
+    case "no-spec":
+      return [
+        {
+          kind: "resolved",
+          spec: {
+            name: "npm",
+            range: "latest",
+          },
+          path: result.path,
+        },
+        options.savePackageManager
+          ? [
+              {
+                code: "no-package-manager-spec",
+                severity: "warning",
+                message:
+                  "No package manager spec found, defaulted to npm latest version. Please set devEngines.packageManager or packageManager in your package.json.",
+                target: NoTarget,
+              },
+            ]
+          : [],
+      ];
+    case "resolved":
+      return [result, []];
+  }
+}
+
+export async function installTypeSpecDependencies(
+  host: CliCompilerHost,
+  options: InstallTypeSpecDependenciesOptions,
+): Promise<readonly Diagnostic[]> {
+  const { directory, stdio = "inherit", savePackageManager } = options;
+  const diagnostics = createDiagnosticCollector();
+  const tracer = createTracer(host.logger).sub("install");
+  try {
+    const { spec, path: packageJsonPath } = diagnostics.pipe(
+      await resolvePackageManagerSpecOrFail(host, tracer, options),
+    );
+    const packageManager = spec.name;
+    const packageManagerConfig = getPackageManagerConfig(packageManager);
+    const manifest = await fetchPackageManifest(packageManager, spec.range);
+    tracer.trace(
+      "fetched-manifest",
+      `Resolved manifest for ${packageManager} at version ${manifest.version}`,
+    );
+
+    const installDir = joinPaths(pmDir, packageManager, manifest.version);
+    await rm(installDir, { recursive: true, force: true });
+    await mkdir(installDir, { recursive: true });
+    const tempDir = await mkdtemp(`tsp-pm-${packageManager}-${manifest.version}`);
+
+    tracer.trace(
+      "downloading-extracting",
+      `Downloading and extracting ${packageManager} at version ${manifest.version} in ${tempDir}`,
+    );
+    const extractResult = await downloadAndExtractPackage(manifest, tempDir, spec.hash?.algorithm);
+    if (spec.hash) {
+      if (spec.hash.value !== extractResult.hash.value) {
+        throw new InstallDependenciesError(
+          `Mismatch hash for package manager. (${spec.hash.algorithm})\n  Expected: ${spec.hash.value}\n  Actual:   ${extractResult.hash}`,
+        );
+      }
+    }
+
+    tracer.trace(
+      "move-temp-to-install",
+      `Move temporary directory ${tempDir} to install directory ${installDir}`,
+    );
+
+    await rename(tempDir, installDir);
```

Comment:
Ah - looks like this is causing the e2e failure - stackoverflow suggests copying/deleting original vs renaming - happens if these are different partitions (might be in our CI)

Reply:
ah, I remember seeing some code to try to create the temp dir always on the same volume, i'll update